### ISSUE TITLE
ViewEncapsulation.Emulated and validator.w3.org html validity results

### ISSUE THREAD

**Issue by fx772** [angular/angular#25355]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior

Without specifying ViewEncapsulation.None in my components, html elements will be rendered with _ng-content-c* tags. The company I work for relies on https://validator.w3.org/ for validity of the webpage. When inserting the source of my html5 angular based webpage, a lot of errors show up.

For example:

`Error: Attribute _ngcontent-c1 not allowed on element header at this point.
(...) <header _ngcontent-c1=""><div  (...)`

I really like the ViewEncapsulation.Emulated(default) functionality, but can't seem to use it to adhere to the W3 rules.

## Expected behavior
Output valid html.
Im not sure if this can be fixed, tags starting with "data-" are allowed and could help with html validity. However, I might be missing a key point or design strategy of the angular devs that made this decision, in that case please let me know.

## Minimal reproduction of the problem with instructions

Compile an Angular (v6.0.7) app and inspect the generated html.

## What is the motivation / use case for changing the behavior?

Output valid html.

## Environment

<pre><code>
Angular version: 6.0.7

Browser:
- [x] Chrome (desktop) version 67.0.3396.99
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.11.3
- Platform:  Windows

Others:

</code></pre>

**Comment by trotyl** [comment]
Duplicate of https://github.com/angular/angular/issues/15891

**Comment by mlc-mlapis** [comment]
@fx772 ... how do you validate it? Just manual copy / paste from run-time maintained DOM via JS? What is the sense of it?

**Comment by fx772** [comment]
My bad.. its a [duplicate question](https://github.com/angular/angular/issues/15891). However, the answers on the OP's question weren't really enlightening enough for me unfortunately.

To put it differently: is there a way to use ViewEncapsulation.Emulated(default) and _not_ see the added tags (during runtime) on elements on which those tags aren't allowed as described in the w3 rules?

@mlc-mlapis 
(procedure of) validation of html on https://validator.w3.org/ by URI, file or direct paste. The sense of it is explained on  [https://validator.w3.org/docs/help.html#validation_basics](https://validator.w3.org/docs/help.html#validation_basics)

Im getting the sense that I might be trying to accomplish something that isn't meant to. No biggy, if it's not possible, we'll use ViewEncapsulation.None and configure our CSS accordingly.

**Comment by mlc-mlapis** [comment]
@fx772 ... then `ViewEncapsulation.ShadowDom` should be the way. It means native solution just based on specs.

**Comment by fx772** [comment]
@mlc-mlapis  The shadowdom option is not sufficient (yet) since we have to support multi browser access.

@robwormald Great stuff thx. Indeed, adding `data-` to the generated tags will get rid of all validation errors as stated earlier. Would be nice if this will be configurable.

**Comment by sagrawal31** [comment]
Agreed with the point of @robwormald. Looks like React is already doing so.

**Comment by avol-io** [comment]
In Italy, but I believe in Europe, all institutional or citizen-designed web applications must follow the rules defined in wcag 2.0.
In particular, there is a European Directive No 19 EU 2016/2102 that regulates it.
In the w3c document that defines https://www.w3.org/TR/WCAG20/ there is the point "4.1.1 Parsing" which is where you have problems.
In particular, this issue does not respect:
 http://www.w3.org/WAI/WCAG20/quickref/#qr-ensure-compat-parses

Basically:
1) It is correct to see html at runtime because screen readers, for example, use html at runtime.
2) it is just a matter of changing _ng prefixes to data-ng - which is a small change.

**Comment by petebacondarwin** [comment]
Making it configurable (or even just switching to CSS classes altogether) would solve this and also #31033.

**Comment by petebacondarwin** [comment]
@avol-io I don't see anything in the WCAG20 requirements that refer to unknown attributes.

> In content implemented using markup languages, elements have complete start and end tags, elements are nested according to their specifications, elements do not contain duplicate attributes, and any IDs are unique, except where the specifications allow these features.
> 
> Note 1: Start and end tags that are missing a critical character in their formation, such as a closing angle bracket or a mismatched attribute value quotation mark are not complete.

Notably it only requires that there are no duplicate attributes and that ID are unique.

So I don't believe that making the generated Angular HTML pass the w3 validator is a requirement for conformance to the WCAG nor the EU directives.

**Comment by avol-io** [comment]
@petebacondarwin 
Few extra info: 
https://www.w3.org/WAI/WCAG21/Techniques/html/H88

And this is error example that came from https://validator.w3.org/nu/ html validator apply the requirement to check html validity
![Schermata 2021-01-26 alle 18 28 06](https://user-images.githubusercontent.com/18249128/105881351-acca3200-6004-11eb-9353-c40656c818c2.png)

**Comment by petebacondarwin** [comment]
The aim, as I understand it from reading https://www.w3.org/WAI/WCAG21/Techniques/html/H88 is to avoid using attributes that the "user agent" is supposed to interpret. By avoiding attributes that are not in the spec, you prevent only a subset of browsers from being able to act on the attribute.  This is not relevant to these attributes added by Angular. They are of no interest to the "user agent" and are not going to impact accessibility for anyone.

**Comment by rooby** [comment]
This issue is not about accessibility, it’s about validity. Accessibility was only brought up in the most recent few comments.

Technically the current WCAG doesn’t require validity (there is an in depth article about the history of this decision at https://www.w3.org/WAI/GL/2005/06/validity-accessibility.html), but that’s beside the point here.

Angular is a widely used and respectable framework and should be setting a good example as other developers will take cues from angular code and copy patterns from it.

It is not setting a good example to say we don’t care about producing valid markup when it is feasible to change it to be valid.

Also, a lot of government & private sector projects I have worked on get run through the validator and questions get asked if there are failures and you say you won’t (can’t) fix them.
While it can generally be smoothed over with an explanation it is not an ideal situation and it is a potential reason for people to prefer another framework that doesn’t output invalid markup. A reason that could be removed by fixing this.
